posthog/tasks/stop_surveys_reached_target.py

```diff
@@ -0,0 +1,50 @@
+from typing import List, Dict
+
+from celery import shared_task
+from django.db.models import UUIDField
+from django.utils import timezone
+
+from posthog.clickhouse.client.connection import Workload
+from posthog.client import sync_execute
+from posthog.models import Survey
+
+
+def _get_surveys_response_counts(surveys_ids: List[UUIDField]) -> Dict[str, int]:
+    data = sync_execute(
+        """
+        SELECT JSONExtractString(properties, '$survey_id') as survey_id, count()
+        FROM events
+        WHERE event = 'survey sent' AND survey_id in %(surveys_ids)s
+        GROUP BY survey_id
+    """,
+        {"surveys_ids": surveys_ids},
+        workload=Workload.OFFLINE,
+    )
+
+    counts = {}
+    for survey_id, count in data:
+        counts[survey_id] = count
+    return counts
+
+
+@shared_task(ignore_result=True)
+def stop_surveys_reached_target() -> None:
+    surveys = Survey.objects.exclude(responses_limit__isnull=True).only("id", "responses_limit")
+    if not surveys:
+        return
+
+    surveys_ids = [survey.id for survey in surveys]
+    response_counts = _get_surveys_response_counts(surveys_ids)
+    for survey in surveys:
+        survey_id = str(survey.id)
+        if survey_id not in response_counts:
+            continue
+
+        assert survey.responses_limit is not None
```

Comment:
I think this assert can fail because it's racy, better not to have it here.

Our operation is roughly idempotent so we don't need to worry about race conditions, other than this.

Reply:
oh yes indeed, good point!

This raises a different problem though that I didn't consider earlier: Race conditions between 2 different celery tasks is no problem, but a race condition between the user and the celery task is problematic.

Consider the case:

1. This task is long running, iterates team by team, so can take ~10 mins (maybe more) to go through the list.
2. A second task updates responses limit to None and ends the survey.
3. 5 mins later, a user comes and decides to restart the survey
4. The first task reaches this team-survey, and stops the survey again

To minimise any chance of this happening, inside `_stop_survey_if_reached_limit` we ought to:

1. `refresh_from_db()`
2. Make sure responses limit is set and > count
3. (?)Make sure end date is not set - not sure if necessary in any case.
4. Then update things.

Haven't thought too hard about this, so let me know if I'm missing something! :)

Reply:
:thinking: I don't follow what you mean by the second and the first task here. If the tasks are executed [once in 12 hours](https://github.com/PostHog/posthog/pull/21528/files#diff-0c209f87d4a097d978f85be965ef58f1676ad11a12eb2eede48830000748f9d4R242) they probably will not overlap, right?

I think the best option here is to set the schedule so that the tasks do not overlap.

Reply:
this is hard to guarantee without an explicit locking mechanism because if tasks have been delayed for a while (because workers are down) they might all run together. So ideally the tasks themselves would be as defensive as possible, and not rely on things going right.

This is a small change that guarantees we won't have future headaches, so I'd say worth making it.

Reply:
Oh, ok, good point. Thanks! Done.

Do you think I can test it without stress testing shenanigans?

Reply:
Yep, similar to here: https://github.com/PostHog/posthog/blob/master/posthog/test/test_feature_flag.py#L5828 - you can call the function (skip going to celery that's going to be hard to test) in multiple threads and add a second function that simulates user update (so just updates the model to start / remove limit / remove end date ) also in the same pool, and check if it fails at least some times without your changes.

Reply:
I dunno... That's what I actually meant by stress testing shenanigans:) Given the simplicity of the implementation + low probability of this event + complexity of the testing approach I would say it's not worth. But I can add it if you think otherwise.

To clarify about the complexity of the testing approach, I actually don't have a vivid picture how I would test this. I understand I need to simulate resuming surveys with a concurrent job running, which is already not too simple. But I also am not sure what to assert here - depending on the execution order the survey might be both stopped and resumed _regardless_ of the `if` that I'm adding. 

WDYT?

Reply:
Happy to not test this bit too, it's pretty overkill, hence didn't ask for it in the first place 👍